distribution/tools/readiness-check-cli/src/main/java/org/elasticsearch/readiness/ReadinessCheckCli.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.readiness;
+
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+
+import org.elasticsearch.cli.Command;
+import org.elasticsearch.cli.SuppressForbidden;
+import org.elasticsearch.cli.Terminal;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.channels.SocketChannel;
+import java.util.Arrays;
+
+public class ReadinessCheckCli extends Command {
+
+    private final OptionSpec<String> hostOption;
+    private final OptionSpec<Integer> portOption;
+
+    public ReadinessCheckCli() {
+        super("A CLI tool to check if Elasticsearch is ready. Exits with return code of 0 if ready.", () -> {});
+        hostOption = parser.acceptsAll(Arrays.asList("h", "host"), "Target hostname").withRequiredArg().defaultsTo("localhost");
+        portOption = parser.acceptsAll(Arrays.asList("p", "port"), "Target port").withRequiredArg().ofType(Integer.class).defaultsTo(9200);
+    }
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options) throws IOException {
+        Integer port = options.valueOf(portOption);
+        String host = options.valueOf(hostOption);
+
+        tryConnect(host, port);
+    }
+
+    @SuppressForbidden(reason = "Intentional socket open. InetSocketAddress is used by the constructor.")
+    void tryConnect(String host, Integer port) throws IOException {
+        try (SocketChannel ignored = SocketChannel.open(new InetSocketAddress(host, port))) {}
```

Comment:
I don't think this is sufficient as a readiness check. The port is opened async to when the node is ready to accept api requests (eg recoveries etc happen). This is why tests wait for green before running.

Reply:
I consulted on this with the distributed team. My initial thought was the same, waiting for green (or yellow) before we claim we are up and running, but that would cause issues on master failover/restart, making the whole cluster appear down. Initially @DaveCTurner suggested we monitor cluster state and wait for the following condition:

```
ClusterChangedEvent.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK) == false
```

Meaning we have sufficiently recovered cluster state to claim the node is up. I believe the ports open after this happens during bootstrap.

After my initial prototype of checking for the cluster state event change, there was a high bandwidth discussion between @tbrooks8, @henningandersen and @DaveCTurner and they recommended this approach going forward.

I suppose we may need to clear-up the exact requirement, are we looking for 'green' node or just node that can respond to requests?

My earlier prototype (of just the server part) that listens on cluster change events is here: https://github.com/grcevski/elasticsearch/tree/feature/readiness_check

Reply:
If a port being open is all we are looking for, then I don't see why a custom tool is needed, it seems possible to do a TCP probe in kubernetes?
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-tcp-liveness-probe

Reply:
Oh, I didn't know you could do TCP probes, I thought it was HTTP and call program only. Then this PR makes no sense. I'll follow-up on the issue with this resolution.